attributes-kt/src/commonMain/kotlin/space/kscience/attributes/PolymorphicAttribute.kt

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2018-2023 KMath contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package space.kscience.attributes
+
+/**
+ * An attribute that has a type parameter for value
+ * @param type parameter-type
+ */
+public abstract class PolymorphicAttribute<T>(public val type: SafeType<T>) : Attribute<T> {
+    override fun equals(other: Any?): Boolean = other != null &&
+            (this::class == other::class) &&
+            (other as? PolymorphicAttribute<*>)?.type == this.type
+
+    override fun hashCode(): Int = this::class.hashCode() + type.hashCode()
+}
+
+
+/**
+ * Get a polymorphic attribute using attribute factory
+ */
+public operator fun <T> Attributes.get(attributeKeyBuilder: () -> PolymorphicAttribute<T>): T? = get(attributeKeyBuilder())
+
+/**
+ * Set a polymorphic attribute using its factory
+ */
+public operator fun <O, T> AttributesBuilder<O>.set(attributeKeyBuilder: () -> PolymorphicAttribute<T>, value: T) {
+    set(attributeKeyBuilder(), value)
+}
```

Comment:
I don't like that the `set` and `get` operators receive lambdas instead of their results, because it looks strange with both brackets and braces:
```kotlin
container[{ ... }]
container[{ ... }] = ...
```
The methods evaluate the lambdas anyway. So I propose simplification of the methods:
```suggestion
/**
 * Get a polymorphic attribute using attribute factory
 */
public operator fun <T> Attributes.get(attributeKey: PolymorphicAttribute<T>): T? = get(attributeKey)

/**
 * Set a polymorphic attribute using its factory
 */
public operator fun <O, T> AttributesBuilder<O>.set(attributeKey: PolymorphicAttribute<T>, value: T) {
    set(attributeKey, value)
}
```